File: src/main/java/org/arbor/gtnn/api/recipe/PlantCasingCondition.java
package org.arbor.gtnn.api.recipe;

import com.google.gson.JsonObject;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.feature.multiblock.IMultiController;
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.api.recipe.RecipeCondition;
import lombok.Getter;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.util.GsonHelper;
import org.arbor.gtnn.block.PlantCasingBlock;
import org.jetbrains.annotations.NotNull;

@Getter
public class PlantCasingCondition extends RecipeCondition {
    public final static PlantCasingCondition INSTANCE = new PlantCasingCondition();

    private PlantCasingBlock plantCasing = PlantCasingBlock.BRONZE;

    public PlantCasingCondition(PlantCasingBlock plantCasing) {
        super();
        this.plantCasing = plantCasing;
    }


    public PlantCasingCondition() {
        super();
    }

    @Override
    public String getType() {
        return "chemical_plant_casing";
    }

    @Override
    public Component getTooltips() {
        return Component.translatable("gtnn.recipe.condition.plant_casing.tooltip", plantCasing.getTier() + 1, plantCasing.getName());
    }

    @Override
    public boolean test(@NotNull GTRecipe recipe, @NotNull RecipeLogic recipeLogic) {
        MetaMachine machine = recipeLogic.getMachine();
        return machine instanceof IMultiController && this.plantCasing != null;
    }

    @Override
    public RecipeCondition createTemplate() {
        return new PlantCasingCondition();
    }

    @NotNull
    @Override
    public JsonObject serialize() {
        JsonObject value = super.serialize();
        value.addProperty("plantCasing", plantCasing.getName());
        return value;
    }

    @Override
    public RecipeCondition deserialize(@NotNull JsonObject config) {
        super.deserialize(config);
        this.plantCasing = PlantCasingBlock.getByName(
                GsonHelper.getAsString(config, "plantCasing", "plantCasing"));
        return this;
    }

    @Override
    public void toNetwork(FriendlyByteBuf buf) {
        super.toNetwork(buf);
        buf.writeUtf(this.plantCasing.getName());
    }

    @Override
    public RecipeCondition fromNetwork(FriendlyByteBuf buf) {
        super.fromNetwork(buf);
        this.plantCasing = PlantCasingBlock.getByNameOrDefault(buf.readUtf());
        return this;
    }
}

File: src/main/kotlin/dev/arbor/gtnn/common/recipe/PlantCasingCondition.kt
package dev.arbor.gtnn.common.recipe

import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic
import com.gregtechceu.gtceu.api.recipe.GTRecipe
import com.gregtechceu.gtceu.api.recipe.RecipeCondition
import com.gregtechceu.gtceu.api.recipe.condition.RecipeConditionType
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.arbor.gtnn.common.machine.multiblock.ChemicalPlantMachine
import dev.arbor.gtnn.data.GTNNRecipeConditions.PLANT_CASING
import net.minecraft.network.chat.Component
import net.minecraft.util.Mth

class PlantCasingCondition : RecipeCondition<PlantCasingCondition> {
    private var tier = 0

    constructor()

    constructor(tier: Int) {
        this.tier = Mth.clamp(tier, 1, 6)
    }

    constructor(isReverse: Boolean, tier: Int) : super(isReverse) {
        this.tier = Mth.clamp(tier, 1, 6)
    }

    override fun getType(): RecipeConditionType<PlantCasingCondition> {
        return PLANT_CASING
    }

    override fun getTooltips(): Component {
        return Component.translatable(
            "gtnn.recipe.condition.plant_casing.tooltip",
            tier, Component.translatable(CASING_TIERS[tier]!!)
        )
    }

    override fun testCondition(gtRecipe: GTRecipe, recipeLogic: RecipeLogic): Boolean {
        val machine = recipeLogic.machine
        if (machine is ChemicalPlantMachine) {
            return machine.casingTier >= tier
        }
        return false
    }

    override fun createTemplate(): PlantCasingCondition {
        return PlantCasingCondition()
    }

    companion object {
        val CODEC: Codec<PlantCasingCondition> = RecordCodecBuilder
            .create { instance: RecordCodecBuilder.Instance<PlantCasingCondition> ->
                isReverse(instance)
                    .and(Codec.INT.fieldOf("plantCasing").forGetter{ it.tier })
                    .apply(instance, ::PlantCasingCondition)
            }

        const val BRONZE: Int = 1
        const val STEEL: Int = 2
        const val ALUMINIUM: Int = 3
        const val STAINLESS_STEEL: Int = 4
        const val TITANIUM: Int = 5
        const val TUNGSTEN_STEEL: Int = 6

        var CASING_TIERS: Map<Int, String> = mapOf(
            BRONZE to "gtnn.recipe.condition.plant_casing.tier.bronze",
            STEEL to "gtnn.recipe.condition.plant_casing.tier.steel",
            ALUMINIUM to "gtnn.recipe.condition.plant_casing.tier.aluminium",
            STAINLESS_STEEL to "gtnn.recipe.condition.plant_casing.tier.stainless_steel",
            TITANIUM to "gtnn.recipe.condition.plant_casing.tier.titanium",
            TUNGSTEN_STEEL to "gtnn.recipe.condition.plant_casing.tier.tungsten_steel"
        )
    }
}
